### DISCUSSION TITLE
Creating external cephobjectstore doesn't create the rgw service

### DISCUSSION THREAD

**Discussion by xakaitetoia** [rook/rook#13863] [Q&A] [answered]
We are using ceph-rook on k8s but with an external ceph cluster. Now everything is working fine for cephfs/rbd/object bucket claims but one thing i have noticed is that the operator hasn't deployed the rgw service to expose port 8080 for my external rgw servers.
 Do I miss anything? I am not deploying a rgw deployment in ceph-rook btw..we are just connecting to external rgw servers e.g ip:8080 which works fine but no rgw service deployed from the operator. 
 
Seems by the end of the gateway settings section https://github.com/rook/rook/blob/master/Documentation/CRDs/Object-Storage/ceph-object-store-crd.md#gateway-settings it says:
```
gateway:
  port: 80
  externalRgwEndpoints:
    - ip: 192.168.39.182
      # hostname: example.com
```
`This will create a service with the endpoint 192.168.39.182 on port 80, pointing to the Ceph object external gateway. All the other settings from the gateway section will be ignored, except for securePort.`

In our case I have the same config but I don't see any rgw service being created by the operator and the endpoint info when you describe the `cephojectstore` is the IP of the host I put and not the rgw service endpoint that it supposed to be created by the operator. 

Do I miss anything? 

Some controller logs with "DEBUG" log:

```
│ 2024-03-04 16:47:36.549314 I | ceph-object-controller: reconciling external object store                                                                                                  │
│ 2024-03-04 16:47:36.549670 D | ceph-object-controller: rgw "rook-ceph/my-store" probe port is {0 8080 }                                                                                   │
│ 2024-03-04 16:47:36.555051 I | ceph-object-controller: creating COSI user "cosi"                                                                                                          │
│ 2024-03-04 16:47:36.614628 I | ceph-spec: created ceph *v1.Secret object "rook-ceph-object-user-my-store-cosi"                                                                            │
│ 2024-03-04 16:47:36.615836 D | ceph-object-controller: domains: +[192.168.58.135]                                                                                                         │
│ 2024-03-04 16:47:36.625155 D | ceph-spec: update event from a CR: "my-store"                                                                                                              │
│ 2024-03-04 16:47:36.626089 D | ceph-spec: update event on CephObjectStore CR                                                                                                              │
│ 2024-03-04 16:47:36.626743 D | clusterdisruption-controller: reconciling "rook-ceph/"                                                                                                     │
│ 2024-03-04 16:47:36.627373 D | ceph-spec: update event from a CR: "my-store"                                                                                                              │
│ 2024-03-04 16:47:36.627845 D | ceph-spec: update event on CephObjectStore CR                                                                                                              │
│ 2024-03-04 16:47:36.628657 D | ceph-object-controller: object store "rook-ceph/my-store" status updated to "Ready"                                                                        │
│ 2024-03-04 16:47:36.629161 D | ceph-object-controller: done reconciling                                                                                                                   │
│ 2024-03-04 16:47:36.629614 D | ceph-object-controller: successfully configured CephObjectStore "rook-ceph/my-store" 
```

**Comment by travisn** [comment]
Do you see any endpoints?
```
kubectl -n rook-ceph get endpoints
```

**Reply by xakaitetoia** [reply]
No unfortunately. I was looking for them too :(

**Reply by parth-gr** [reply]
@xakaitetoia can you share the output of `kubeclt get cephobjectstore -o yaml`

**Reply by xakaitetoia** [reply]
Sure:
```
apiVersion: v1
items:
- apiVersion: ceph.rook.io/v1
  kind: CephObjectStore
  metadata:
    creationTimestamp: "2024-03-04T16:47:36Z"
    finalizers:
    - cephobjectstore.ceph.rook.io
    generation: 2
    name: my-store
    namespace: rook-ceph
    resourceVersion: "406715"
    uid: f359e398-87e6-49ba-82c3-be9cab094a8f
  spec:
    dataPool:
      erasureCoded:
        codingChunks: 0
        dataChunks: 0
      mirroring: {}
      quotas: {}
      replicated:
        size: 0
      statusCheck:
        mirror: {}
    gateway:
      externalRgwEndpoints:
      - ip: 192.168.58.135
      placement: {}
      port: 8080
      resources: {}
    healthCheck: {}
    metadataPool:
      erasureCoded:
        codingChunks: 0
        dataChunks: 0
      mirroring: {}
      quotas: {}
      replicated:
        size: 0
      statusCheck:
        mirror: {}
    zone:
      name: ""
  status:
    endpoints:
      insecure:
      - http://192.168.58.135:8080
      secure: []
    info:
      endpoint: http://192.168.58.135:8080
    observedGeneration: 2
    phase: Ready
kind: List
metadata:
  resourceVersion:  ""

**Comment by BlaineEXE** [comment, marked as answer]
Rook was modified to no longer create the Service (but it leaves old, legacy Services in place). It seems that the documentation may need to be updated to reflect that.

We got rid of the service because users reported certificate issues when Rook created Services for RGWs that had TLS (HTTPS) enabled. It is now up to the user to create a Service if they want to use one. 

We recommend using the endpoint as you have specified it in the `externalRgwEndpoints` list and not bothering with a Service. If that isn't possible for whatever reason, you can create an external service, and then it is best to put only that service's FQDN into the `externalRgwEndpoints` list.

And as a note: For the external RGW, if the IP given isn't a guaranteed static IP, I would recommend using an FQDN instead.

**Reply by xakaitetoia** [reply]
Thanks a lot @BlaineEXE, yeah this is how we have performed a workaround for now. I really appreciate your input and yeah maybe the documentation needs to be adjusted on this point.
